creusot/src/backend/clone_map.rs

```diff
@@ -324,89 +281,41 @@ impl<'tcx> CloneNames<'tcx> {
-                                            Symbol::intern(&("T_".to_string() + name.as_str()))
-                                        };
-                                    let fresh = self.counts.freshen(name);
-                                    self.adt_names.insert(did, fresh);
-                                    fresh
-                                }
-                            }
-                        };
-                        let alias = Ident::build(alias.as_str());
-                        let modl = ModulePath::new(self.tcx, did, util::NS::T);
-                        let name = Symbol::intern(&*item_name(self.tcx, did, Namespace::TypeNS));
-                        Kind::Used(modl, alias, name)
-                    }
+        self.names.entry(key).or_insert_with(|| {
+            if let Some((did, _)) = key.did()
+                && let Some(why3_modl) = get_builtin(self.tcx, did)
+            {
+                let qname = QName::from_string(why3_modl.as_str());
+                let name = qname.name.clone();
+                if let Some(modl) = qname.module_qname() {
+                    return Kind::UsedBuiltin(modl, Symbol::intern(&*name));
                 } else {
-                    Kind::Named(Symbol::intern("hidden_type_name"))
+                    return Kind::Named(Symbol::intern(&*name));
                 }
             }
-            Dependency::Item(id, _) if util::item_type(self.tcx, id) == ItemType::Variant => {
-                // This branch doesn't seem to be used
-                let ty = self.tcx.parent(id);
-                let modl = ModulePath::new(self.tcx, ty, util::NS::T);
-                let alias = modl.why3_ident();
-                Kind::Used(modl, alias, key.base_ident(self.tcx).unwrap())
-            }
-            _ => {
-                if let Dependency::Item(id, _) = key
-                    && let Some(why3_modl) = contracts_items::get_builtin(self.tcx, id)
-                {
-                    let qname = QName::from_string(why3_modl.as_str());
-                    let name = qname.name.clone();
-                    let modl = qname.module_qname();
-                    return Kind::UsedBuiltin(modl, Symbol::intern(&*name));
-                };
-
-                key.base_ident(self.tcx)
-                    .map_or(Kind::Unnamed, |base| Kind::Named(self.counts.freshen(base)))
-            }
+            key.base_ident(self.tcx)
+                .map_or(Kind::Unnamed, |base| Kind::Named(self.counts.freshen(base)))
         })
     }
 }
 
 impl NameSupply {
     pub(crate) fn freshen(&mut self, sym: Symbol) -> Symbol {
         let count: usize = *self.name_counts.entry(sym).and_modify(|c| *c += 1).or_insert(0);
+        // FIXME: if we don't do use the initial ident when count == 0, then the ident clashes
+        // with local variables
+        /*if count == 0 {
+            sym
+        } else {*/
         Symbol::intern(&format!("{sym}'{count}"))
+        /*}*/
```

Comment:
what's the status of this

Reply:
I think we should change the design of name handling in Creusot. Here is my thoughts:
- The `Ident` type of the Why3 crate should carry a non-freshened (but mangled) string, and a unique identifier (e.g., u64).
- `Ident::new` should generate a fresh u64, and the creusot crate should carry around the same `Ident` when the refer to one variable in the same scope.
- When printing, the Why3 crate should track the content of the environment, and freshen identifiers if there are some clashes.

I already discussed with @Lysxia about all that. He said he would work on this, but we are waiting for this PR because his change would conflict a lot.